File: greenfloor/adapters/dexie.py
from __future__ import annotations

import json
import time
import urllib.error
import urllib.parse
import urllib.request
from dataclasses import dataclass, field
from typing import Any


@dataclass
class _RowCache:
    """Minimal TTL cache for a list of dict rows."""

    ttl: int
    _rows: list[dict] | None = field(default=None, init=False, repr=False)
    _cached_at: float | None = field(default=None, init=False, repr=False)

    def get_if_fresh(self, now: float) -> list[dict] | None:
        if (
            self._rows is not None
            and self._cached_at is not None
            and (now - self._cached_at) <= self.ttl
        ):
            return list(self._rows)
        return None

    def store(self, rows: list[dict], now: float) -> list[dict]:
        self._rows = list(rows)
        self._cached_at = now
        return list(rows)

    def stale(self) -> list[dict]:
        return list(self._rows) if self._rows is not None else []


class DexieAdapter:
    def __init__(
        self,
        base_url: str,
        *,
        cache_ttl_seconds: int = 900,
    ) -> None:
        self.base_url = base_url.rstrip("/")
        ttl = max(1, int(cache_ttl_seconds))
        self._token_cache = _RowCache(ttl=ttl)
        self._ticker_cache = _RowCache(ttl=ttl)

    def get_tokens(self) -> list[dict]:
        url = f"{self.base_url}/v1/swap/tokens"
        with urllib.request.urlopen(url, timeout=15) as resp:
            payload = json.loads(resp.read().decode("utf-8"))
        tokens = payload.get("tokens", payload)
        if isinstance(tokens, list):
            return [t for t in tokens if isinstance(t, dict)]
        return []

    def get_offers(self, offered: str, requested: str) -> list[dict]:
        q = urllib.parse.urlencode({"offered": offered, "requested": requested})
        url = f"{self.base_url}/v1/offers?{q}"
        with urllib.request.urlopen(url, timeout=20) as resp:
            payload = json.loads(resp.read().decode("utf-8"))
        offers = payload.get("offers", [])
        return [o for o in offers if isinstance(o, dict)]

    def get_offer(self, offer_id: str, *, timeout: int = 20) -> dict[str, Any]:
        clean_offer_id = str(offer_id).strip()
        if not clean_offer_id:
            raise ValueError("offer_id is required")
        url = f"{self.base_url}/v1/offers/{urllib.parse.quote(clean_offer_id)}"
        with urllib.request.urlopen(url, timeout=timeout) as resp:
            payload = json.loads(resp.read().decode("utf-8"))
        if isinstance(payload, dict):
            return payload
        return {"success": False, "error": "invalid_response_format"}

    def post_offer(
        self,
        offer: str,
        *,
        drop_only: bool = True,
        claim_rewards: bool | None = None,
    ) -> dict[str, Any]:
        payload: dict[str, Any] = {
            "offer": offer,
            "drop_only": bool(drop_only),
        }
        if claim_rewards is not None:
            payload["claim_rewards"] = bool(claim_rewards)
        url = f"{self.base_url}/v1/offers"
        body = json.dumps(payload).encode("utf-8")
        req = urllib.request.Request(
            url,
            data=body,
            method="POST",
            headers={"Content-Type": "application/json"},
        )
        try:
            with urllib.request.urlopen(req, timeout=20) as resp:
                result = json.loads(resp.read().decode("utf-8"))
        except urllib.error.HTTPError as exc:
            raw = exc.read().decode("utf-8", errors="replace").strip()
            snippet = raw[:500] if raw else ""
            error = f"dexie_http_error:{exc.code}"
            if snippet:
                error = f"{error}:{snippet}"
            return {"success": False, "error": error}
        except urllib.error.URLError as exc:
            return {"success": False, "error": f"dexie_network_error:{exc.reason}"}
        if isinstance(result, dict):
            return result
        return {"success": False, "error": "invalid_response_format"}

    def cancel_offer(self, offer_id: str) -> dict[str, Any]:
        clean_offer_id = offer_id.strip()
        url = f"{self.base_url}/v1/offers/{urllib.parse.quote(clean_offer_id)}/cancel"
        body = json.dumps({"id": clean_offer_id}).encode("utf-8")
        req = urllib.request.Request(
            url,
            data=body,
            method="POST",
            headers={"Content-Type": "application/json"},
        )
        with urllib.request.urlopen(req, timeout=20) as resp:
            result = json.loads(resp.read().decode("utf-8"))
        if isinstance(result, dict):
            return result
        return {"success": False, "error": "invalid_response_format"}

    def lookup_token_by_cat_id(self, cat_id_hex: str) -> dict | None:
        """Find a token by CAT asset ID across swap tokens and v3 tickers."""
        target = cat_id_hex.strip().lower()
        if not target:
            return None

        for row in self._fetch_token_rows():
            if _row_matches_cat_target(row, target):
                return row

        ticker_rows = self._fetch_ticker_rows()
        for row in ticker_rows:
            if _row_matches_cat_target(row, target, include_ticker_split=True):
                return row
        return None

    def lookup_token_by_symbol(
        self,
        symbol: str,
        *,
        label_matcher: Any | None = None,
    ) -> dict | None:
        """Find a token by symbol/name/code with optional fuzzy label matching."""
        target = symbol.strip()
        if not target:
            return None
        match_fn = label_matcher or _case_insensitive_match
        for row in self._fetch_token_rows():
            for key in ("code", "name", "id"):
                if match_fn(str(row.get(key, "")), target):
                    return row
        return None

    def _cached_fetch(self, cache: _RowCache, fetcher: Any) -> list[dict]:
        """Fetch rows through *cache*, falling back to stale rows on error."""
        now = time.time()
        fresh = cache.get_if_fresh(now)
        if fresh is not None:
            return fresh
        try:
            rows = fetcher()
        except Exception:
            return cache.stale()
        return cache.store(rows, now)

    def _fetch_token_rows(self) -> list[dict]:
        return self._cached_fetch(self._token_cache, self.get_tokens)

    def _fetch_ticker_rows(self) -> list[dict]:
        def _fetch() -> list[dict]:
            url = f"{self.base_url}/v3/prices/tickers"
            with urllib.request.urlopen(url, timeout=20) as resp:
                payload = json.loads(resp.read().decode("utf-8"))
            if isinstance(payload, list):
                return [r for r in payload if isinstance(r, dict)]
            if isinstance(payload, dict):
                tickers = payload.get("tickers")
                if isinstance(tickers, list):
                    return [r for r in tickers if isinstance(r, dict)]
            return []

        return self._cached_fetch(self._ticker_cache, _fetch)


def _row_matches_cat_target(row: dict, target: str, *, include_ticker_split: bool = False) -> bool:
    candidates = {
        str(row.get("assetId", "")).strip().lower(),
        str(row.get("asset_id", "")).strip().lower(),
        str(row.get("id", "")).strip().lower(),
        str(row.get("tokenId", "")).strip().lower(),
        str(row.get("token_id", "")).strip().lower(),
        str(row.get("base_currency", "")).strip().lower(),
        str(row.get("target_currency", "")).strip().lower(),
    }
    ticker_id = str(row.get("ticker_id", "")).strip().lower()
    if ticker_id:
        candidates.add(ticker_id)
        if include_ticker_split and "_" in ticker_id:
            base, quote = ticker_id.split("_", 1)
            candidates.add(base)
            candidates.add(quote)
    return target in candidates


def _case_insensitive_match(left: str, right: str) -> bool:
    a = left.strip().lower()
    b = right.strip().lower()
    return bool(a and b and a == b)

File: greenfloor-engine/src/adapters/dexie.rs
use serde_json::{json, Value};

use crate::cycle::{
    dexie_invalid_offer_retry_sleep, dexie_invalid_offer_should_retry,
    is_transient_dexie_visibility_404_error,
};
use crate::error::{SignerError, SignerResult};
use crate::offer::publish::dexie_offer_asset_expectation_error;

const INVALID_OFFER_RETRY_MAX_ATTEMPTS: u32 = 4;
const INVALID_OFFER_RETRY_INITIAL_SLEEP_SECONDS: f64 = 1.0;
const VISIBILITY_POST_MAX_ATTEMPTS: u32 = 3;
const VISIBILITY_POST_DELAY_SECONDS: f64 = 2.0;

#[derive(Debug, Clone)]
pub struct DexieClient {
    base_url: String,
    http: reqwest::Client,
}

impl DexieClient {
    pub fn new(base_url: impl Into<String>) -> Self {
        Self {
            base_url: base_url.into().trim_end_matches('/').to_string(),
            http: reqwest::Client::new(),
        }
    }

    #[must_use]
    pub fn base_url(&self) -> &str {
        &self.base_url
    }

    /// Post offer.
    ///
    /// # Errors
    ///
    /// Returns an error if the operation fails.
    pub async fn post_offer(
        &self,
        offer: &str,
        drop_only: bool,
        claim_rewards: bool,
    ) -> SignerResult<Value> {
        let payload = json!({
            "offer": offer,
            "drop_only": drop_only,
            "claim_rewards": claim_rewards,
        });
        let url = format!("{}/v1/offers", self.base_url);
        let response = self
            .http
            .post(url)
            .json(&payload)
            .timeout(std::time::Duration::from_secs(20))
            .send()
            .await
            .map_err(|err| SignerError::Other(format!("dexie_network_error:{err}")))?;
        Self::parse_response(response).await
    }

    /// Get offer.
    ///
    /// # Errors
    ///
    /// Returns an error if the operation fails.
    pub async fn get_offer(&self, offer_id: &str) -> SignerResult<Value> {
        let clean_offer_id = offer_id.trim();
        if clean_offer_id.is_empty() {
            return Err(SignerError::Other("offer_id is required".to_string()));
        }
        let encoded = urlencoding::encode(clean_offer_id);
        let url = format!("{}/v1/offers/{encoded}", self.base_url);
        let response = self
            .http
            .get(url)
            .timeout(std::time::Duration::from_secs(20))
            .send()
            .await
            .map_err(|err| SignerError::Other(format!("dexie_get_offer_error:{err}")))?;
        Self::parse_response(response).await
    }

    /// Get offers.
    ///
    /// # Errors
    ///
    /// Returns an error if the operation fails.
    pub async fn get_offers(&self, offered: &str, requested: &str) -> SignerResult<Vec<Value>> {
        let query = format!(
            "offered={}&requested={}",
            urlencoding::encode(offered.trim()),
            urlencoding::encode(requested.trim())
        );
        let url = format!("{}/v1/offers?{query}", self.base_url);
        let response = self
            .http
            .get(url)
            .timeout(std::time::Duration::from_secs(20))
            .send()
            .await
            .map_err(|err| SignerError::Other(format!("dexie_get_offers_error:{err}")))?;
        let payload = Self::parse_response(response).await?;
        let offers = payload
            .get("offers")
            .and_then(Value::as_array)
            .cloned()
            .unwrap_or_default();
        Ok(offers)
    }

    /// Get swap tokens.
    ///
    /// # Errors
    ///
    /// Returns an error if the operation fails.
    pub async fn get_swap_tokens(&self) -> SignerResult<Vec<Value>> {
        let url = format!("{}/v1/swap/tokens", self.base_url);
        let response = self
            .http
            .get(url)
            .timeout(std::time::Duration::from_secs(15))
            .send()
            .await
            .map_err(|err| SignerError::Other(format!("dexie_get_tokens_error:{err}")))?;
        let payload = Self::parse_response(response).await?;
        let tokens = payload
            .get("tokens")
            .and_then(Value::as_array)
            .cloned()
            .unwrap_or_else(|| payload.as_array().cloned().unwrap_or_default());
        Ok(tokens
            .into_iter()
            .filter(serde_json::Value::is_object)
            .collect())
    }

    /// Get price tickers.
    ///
    /// # Errors
    ///
    /// Returns an error if the operation fails.
    pub async fn get_price_tickers(&self) -> SignerResult<Vec<Value>> {
        let url = format!("{}/v3/prices/tickers", self.base_url);
        let response = self
            .http
            .get(url)
            .timeout(std::time::Duration::from_secs(20))
            .send()
            .await
            .map_err(|err| SignerError::Other(format!("dexie_get_tickers_error:{err}")))?;
        let payload = Self::parse_response(response).await?;
        if let Some(rows) = payload.as_array() {
            return Ok(rows.iter().filter(|row| row.is_object()).cloned().collect());
        }
        Ok(payload
            .get("tickers")
            .and_then(Value::as_array)
            .cloned()
            .unwrap_or_default()
            .into_iter()
            .filter(serde_json::Value::is_object)
            .collect())
    }

    /// Lookup token by cat id.
    ///
    /// # Errors
    ///
    /// Returns an error if the operation fails.
    pub async fn lookup_token_by_cat_id(&self, cat_id_hex: &str) -> SignerResult<Option<Value>> {
        let target = cat_id_hex.trim().to_ascii_lowercase();
        if target.is_empty() {
            return Ok(None);
        }
        for row in self.get_swap_tokens().await? {
            if row_matches_cat_target(&row, &target, false) {
                return Ok(Some(row));
            }
        }
        for row in self.get_price_tickers().await? {
            if row_matches_cat_target(&row, &target, true) {
                return Ok(Some(row));
            }
        }
        Ok(None)
    }

    /// Lookup token by symbol.
    ///
    /// # Errors
    ///
    /// Returns an error if the operation fails.
    pub async fn lookup_token_by_symbol(&self, symbol: &str) -> SignerResult<Option<Value>> {
        let target = symbol.trim();
        if target.is_empty() {
            return Ok(None);
        }
        for row in self.get_swap_tokens().await? {
            for key in ["code", "name", "id"] {
                if case_insensitive_match(
                    row.get(key).and_then(Value::as_str).unwrap_or(""),
                    target,
                ) {
                    return Ok(Some(row));
                }
            }
        }
        Ok(None)
    }

    /// Cancel offer.
    ///
    /// # Errors
    ///
    /// Returns an error if the operation fails.
    pub async fn cancel_offer(&self, offer_id: &str) -> SignerResult<Value> {
        let clean_offer_id = offer_id.trim();
        if clean_offer_id.is_empty() {
            return Err(SignerError::Other("offer_id is required".to_string()));
        }
        let encoded = urlencoding::encode(clean_offer_id);
        let url = format!("{}/v1/offers/{encoded}/cancel", self.base_url);
        let response = self
            .http
            .post(url)
            .json(&json!({"id": clean_offer_id}))
            .timeout(std::time::Duration::from_secs(20))
            .send()
            .await
            .map_err(|err| SignerError::Other(format!("dexie_cancel_offer_error:{err}")))?;
        Self::parse_response(response).await
    }

    async fn parse_response(response: reqwest::Response) -> SignerResult<Value> {
        let status = response.status();
        let body = response
            .text()
            .await
            .map_err(|err| SignerError::Other(format!("dexie_read_error:{err}")))?;
        if !status.is_success() {
            let snippet: String = body.chars().take(500).collect();
            let error = if snippet.is_empty() {
                format!("dexie_http_error:{}", status.as_u16())
            } else {
                format!("dexie_http_error:{}:{snippet}", status.as_u16())
            };
            return Ok(json!({"success": false, "error": error}));
        }
        serde_json::from_str(&body)
            .map_err(|err| SignerError::Other(format!("dexie_json_error:{err}")))
    }
}

fn case_insensitive_match(left: &str, right: &str) -> bool {
    let a = left.trim().to_ascii_lowercase();
    let b = right.trim().to_ascii_lowercase();
    !a.is_empty() && a == b
}

fn row_matches_cat_target(row: &Value, target: &str, include_ticker_split: bool) -> bool {
    let mut candidates = std::collections::HashSet::new();
    for key in [
        "assetId",
        "asset_id",
        "id",
        "tokenId",
        "token_id",
        "base_currency",
        "target_currency",
    ] {
        if let Some(value) = row.get(key).and_then(Value::as_str) {
            let trimmed = value.trim().to_ascii_lowercase();
            if !trimmed.is_empty() {
                candidates.insert(trimmed);
            }
        }
    }
    if let Some(ticker_id) = row.get("ticker_id").and_then(Value::as_str) {
        let ticker_id = ticker_id.trim().to_ascii_lowercase();
        if !ticker_id.is_empty() {
            candidates.insert(ticker_id.clone());
            if include_ticker_split {
                if let Some((base, quote)) = ticker_id.split_once('_') {
                    candidates.insert(base.to_string());
                    candidates.insert(quote.to_string());
                }
            }
        }
    }
    candidates.contains(target)
}

#[must_use]
pub fn dexie_offer_view_url(dexie_base_url: &str, offer_id: &str) -> String {
    let clean_offer_id = offer_id.trim();
    if clean_offer_id.is_empty() {
        return String::new();
    }
    let trimmed = dexie_base_url.trim();
    let host = trimmed
        .trim_start_matches("https://")
        .trim_start_matches("http://")
        .trim_end_matches('/');
    let host = if let Some(rest) = host.strip_prefix("api-testnet.") {
        format!("testnet.{rest}")
    } else if let Some(rest) = host.strip_prefix("api.") {
        rest.to_string()
    } else {
        host.to_string()
    };
    format!(
        "https://{host}/offers/{}",
        urlencoding::encode(clean_offer_id)
    )
}

/// Post dexie offer with invalid offer retry.
///
/// # Errors
///
/// Returns an error if the operation fails.
pub async fn post_dexie_offer_with_invalid_offer_retry(
    dexie: &DexieClient,
    offer_text: &str,
    drop_only: bool,
    claim_rewards: bool,
) -> SignerResult<Value> {
    let mut attempt = 0u32;
    loop {
        let result = dexie
            .post_offer(offer_text, drop_only, claim_rewards)
            .await?;
        let error = result
            .get("error")
            .and_then(Value::as_str)
            .unwrap_or("")
            .trim()
            .to_string();
        if !dexie_invalid_offer_should_retry(&error, attempt, INVALID_OFFER_RETRY_MAX_ATTEMPTS) {
            return Ok(result);
        }
        let sleep_seconds =
            dexie_invalid_offer_retry_sleep(attempt, INVALID_OFFER_RETRY_INITIAL_SLEEP_SECONDS);
        tokio::time::sleep(std::time::Duration::from_secs_f64(sleep_seconds)).await;
        attempt += 1;
    }
}

pub async fn verify_dexie_offer_visible_by_id(
    dexie: &DexieClient,
    offer_id: &str,
    expected_offered_asset_id: &str,
    expected_offered_symbol: &str,
    expected_requested_asset_id: &str,
    expected_requested_symbol: &str,
) -> Option<String> {
    let clean_offer_id = offer_id.trim();
    if clean_offer_id.is_empty() {
        return Some("dexie_offer_missing_id_after_publish".to_string());
    }
    let attempts = 4usize;
    let delay_seconds = 1.5;
    let mut last_error = "dexie_offer_not_visible_after_publish".to_string();
    for attempt in 1..=attempts {
        let payload = match dexie.get_offer(clean_offer_id).await {
            Ok(payload) => payload,
            Err(err) => {
                last_error = format!("dexie_get_offer_error:{err}");
                if attempt < attempts {
                    tokio::time::sleep(std::time::Duration::from_secs_f64(delay_seconds)).await;
                }
                continue;
            }
        };
        let offer_payload = payload.get("offer");
        let visible_id = offer_payload
            .and_then(Value::as_object)
            .and_then(|obj| obj.get("id"))
            .and_then(Value::as_str)
            .unwrap_or("")
            .trim();
        if visible_id == clean_offer_id {
            if let Some(offer_obj) = offer_payload.and_then(Value::as_object) {
                if let Some(asset_error) = dexie_offer_asset_expectation_error(
                    offer_obj.get("offered").unwrap_or(&Value::Null),
                    offer_obj.get("requested").unwrap_or(&Value::Null),
                    expected_offered_asset_id,
                    expected_offered_symbol,
                    expected_requested_asset_id,
                    expected_requested_symbol,
                ) {
                    return Some(asset_error);
                }
            }
            return None;
        }
        last_error = "dexie_offer_visibility_payload_mismatch".to_string();
        if attempt < attempts {
            tokio::time::sleep(std::time::Duration::from_secs_f64(delay_seconds)).await;
        }
    }
    Some(last_error)
}

#[derive(Debug, Clone)]
pub struct PostOfferPhaseDexieParams<'a> {
    pub dexie: &'a DexieClient,
    pub offer_text: &'a str,
    pub drop_only: bool,
    pub claim_rewards: bool,
    pub expected_offered_asset_id: &'a str,
    pub expected_offered_symbol: &'a str,
    pub expected_requested_asset_id: &'a str,
    pub expected_requested_symbol: &'a str,
}

/// Post offer phase dexie.
///
/// # Errors
///
/// Returns an error if the operation fails.
pub async fn post_offer_phase_dexie(params: PostOfferPhaseDexieParams<'_>) -> SignerResult<Value> {
    let PostOfferPhaseDexieParams {
        dexie,
        offer_text,
        drop_only,
        claim_rewards,
        expected_offered_asset_id,
        expected_offered_symbol,
        expected_requested_asset_id,
        expected_requested_symbol,
    } = params;
    let mut last_result =
        json!({"success": false, "error": "dexie_offer_not_visible_after_publish"});
    let mut last_visibility_error = String::new();
    for attempt in 1..=VISIBILITY_POST_MAX_ATTEMPTS {
        let result =
            post_dexie_offer_with_invalid_offer_retry(dexie, offer_text, drop_only, claim_rewards)
                .await?;
        last_result = result.clone();
        if result.get("success").and_then(Value::as_bool) != Some(true) {
            return Ok(result);
        }
        let posted_offer_id = result
            .get("id")
            .and_then(Value::as_str)
            .unwrap_or("")
            .trim()
            .to_string();
        if let Some(visibility_error) = verify_dexie_offer_visible_by_id(
            dexie,
            &posted_offer_id,
            expected_offered_asset_id,
            expected_offered_symbol,
            expected_requested_asset_id,
            expected_requested_symbol,
        )
        .await
        {
            last_visibility_error = visibility_error;
            if !is_transient_dexie_visibility_404_error(&last_visibility_error) {
                let mut failed = result;
                if let Value::Object(obj) = &mut failed {
                    obj.insert("success".to_string(), Value::Bool(false));
                    obj.insert(
                        "error".to_string(),
                        Value::String(last_visibility_error.clone()),
                    );
                }
                return Ok(failed);
            }
            if attempt < VISIBILITY_POST_MAX_ATTEMPTS {
                tokio::time::sleep(std::time::Duration::from_secs_f64(
                    VISIBILITY_POST_DELAY_SECONDS,
                ))
                .await;
            }
            continue;
        }
        return Ok(result);
    }
    let mut failed = last_result;
    if let Value::Object(obj) = &mut failed {
        obj.insert("success".to_string(), Value::Bool(false));
        obj.insert(
            "error".to_string(),
            Value::String(if last_visibility_error.is_empty() {
                "dexie_offer_not_visible_after_publish".to_string()
            } else {
                last_visibility_error
            }),
        );
    }
    Ok(failed)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn dexie_view_url_strips_api_prefix() {
        assert_eq!(
            dexie_offer_view_url("https://api.dexie.space", "offer-123"),
            "https://dexie.space/offers/offer-123"
        );
        assert_eq!(
            dexie_offer_view_url("https://api-testnet.dexie.space", "offer-123"),
            "https://testnet.dexie.space/offers/offer-123"
        );
    }
}
